libc/watcom/syscall/crt0.c

```diff
@@ -46,7 +49,7 @@ noreturn void exit(int status)
  *
  * main(ac, av) called by AX, DX (small/medium) or AX, CX:BX (compact/large) model
  */
-#pragma aux main "*" modify [ bx cx dx si di ]
+#pragma aux (elkscall) main "*"
```

Comment:
The reason for this pragma aux for `main` is actually to keep OWC from automatically renaming the symbol with an underscore and force a `_cdecl` naming convention, rather than an ELKS calling sequence of some kind. The same pragma is used in elks/include/arch/cdefs.h, which with this change now does not match.

The second reason for this pragma is to inhibit register saves in main, but frankly that probably doesn't matter as much, except that its easy to do in the same pragma.

Controlling the process by which `main` is called is important in crt0.s due to the way that we have had to rewrite the argc/argv vector (passed with 2-byte pointers by ELKS) to 4-byte pointers in compact and large models. I have tried to design the whole process so that needless stack pushing or register saves are performed, and instead use the very nice OpenWatcom register calling model, which I am finding is quite cool compared to stack-calling convention(s). (The syscall wrappers are a good example of this, where very little code is generated to convert OWC syscall wrappers to an ELKS system call. This eliminated lots of code that ELKS uses in the ia16-elf-gcc cdecl default calling convention, linked into every program). Overall, I am extremely impressed with the OWC register allocations produced, they are quite different than ia16-elf-gcc, even though in some cases quite a few SS: overrides are seen (but with big benefit allowing DS to be scratch at times).

Reply:
the main function is defined by C standard and is entry point to C standard environment that it should be handled carefully.
Thanks #pragma aux you can change it completely but caller in startup code must follow you definition of main.
if you want use cdecl calling convention in startup code then you can change calling convention of main by #pragma aux to cdecl standard way as below.
`#pragma aux (__cdecl) main`
It do main cdecl function with all necessary changes parameters on stack name without mangling, changes preserved registers etc.

Reply:
I looked into crt0 code and it is a little strange because you change standard C function calling convention that it will have a problems with OW compiler.
What sense for such changes if crt0 is specific for OW that can use OW standard behaviour as main calling convention etc without any change of prototype and #pragma aux?
All C function can use OW default calling convention and not need any hack.

Reply:
This is one of the early changes made when trying to optimize function calling to not save registers when not needed, along with some changes to main's signature. They are no longer needed, and now I agree this is over-optimization and main should not have and does not need hack to not save registers and should be treated as normal calling convention. Approved, but please test that elkscmd/basic still compiles and runs:
```
ewcc basic.c
ewcc host.c
ewcc host-stubs.c
ewlink basic.obj host.obj host-stubs.obj
```

Please delete special calling convention for main in both crt0.c as well as include/arch/cdefs.h.